SwiftUI Example Project

amyoulton opened the pull request:
** This is an ongoing draft PR for collaboration/feedback as we continue to build out the project **

The new example project is being built out with SwiftUI. I have set the project settings and built out a mock "API" using `Swifter`. 

I originally wanted to included live previews within the application, but because the network is being run within the application, if you have a build open, the live preview throws a fatal error because a port is already open. That's the nature of the network being within the app. Since it's a super fast app to build, so I've chosen to keep live previews off.

## The Project

Currently, I've added the following things into the new SwiftUI project:

### The "API"

This includes my JSON data that I will continue to add to, and my network that leverages Swifter. I have currently added JSON data for:

- posts
- version

I'll update the above list as the project continues getting built out, as more JSON will be added.

### The Build

- HomeView & HomeVM (containing posts and a version print)
- Post Repo & Version Repo, to showcase different Netable Requests. Version serves to be the most "basic" form of a request from start to finish, so despite it being small I wanted to keep it seperate. Posts will likely keep expanding with more request options as I continue.
- PostRequest & Version Request. Similar to above, Post will likely keep expanding to showcase using other features.

amyoulton commented on Netable/Example/ExampleApp.swift:
```diff
@@ -0,0 +1,22 @@
+@main
+struct ExampleApp: App {
+    init() {
+        ExampleNetworkService.shared
```
This is likely *not* an ideal way to render this network out, but is serving as a simple way to do it for right now while building. This can likely be improved upon before completing the project.

amyoulton replied:
This network calling can be improved upon, it's just a simple way to get it going now.

brendanlensink commented on Netable/Example/ViewModels/UserVM.swift:
```diff
@@ -0,0 +1,20 @@
+class UserVM: ObservableObject {
+    @Published var user: User?
+
+
```
It might be worth either adding swiftlint to this project temporarily before we merge - I've noticed a handful of these little polish-y typos

brendanlensink commented on Netable/Example/Services/GraphQLNetworkService.swift:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    func updatePost(title: String, content: String) {
+        Task {
```
We may want to switch this so to `updatePost() async throws` to propagate handling async/throws up to the view model layer to match the other services

amyoulton commented on Netable/Example/Views/CreateGraphQLPostView.swift:
```diff
@@ -0,0 +1,46 @@
+struct CreateGraphQLPostView: View {
+    @Environment(\.presentationMode) var presentationMode
+    @ObservedObject var viewModel: GraphQLVM
+
```
I got a little carried away with the SwiftUI here but the form was basically invisible without it 🙃

brendanlensink commented on Netable/Example/Models/Version.swift:
```diff
@@ -0,0 +1,18 @@
+
+import Foundation
+
+struct SimpleVersion: Decodable {
```
I think this is backwards, SimpleVersion has more fields than Version 😅

brendanlensink commented on Netable/Example/Requests/GraphQL/UpdatePostsMutation.swift:
```diff
@@ -0,0 +1,33 @@
+
+    var input: UpdatePostMutationInput
+
+    var source = GraphQLQuerySource.resource("UpdatePostsMutation")
```
Might be worth demoing one of the other ways of providing a GraphQL source here

brendanlensink commented on Netable/Example/Services/AuthNetworkService.swift:
```diff
@@ -0,0 +1,91 @@
+            baseURL: URL(string: "http://localhost:8080/")!)
+
+        unauthNetable.requestFailurePublisher.sink { [weak self] error in
+            guard case let NetableError.httpError(statusCode, _) = error, statusCode == 401 else {
```
nitpicky, but you can use a filter here instead of returning early to clean it up a little

amyoulton replied:
It's an enum and not an array, so instead of using `filter` I just did a simple `if` statement that if it doesn't pass nothing gets sent. If there is a better way to handle this I'd love to know how to improve it!

brendanlensink commented on Netable/Example/Services/AuthNetworkService.swift:
```diff
@@ -0,0 +1,91 @@
+            requestFailureDelegate: ErrorService.shared)
+    }
+
+    func getUser() async throws {
```
can remove the async throws here I think

brendanlensink commented on Netable/Example/Services/AuthNetworkService.swift:
```diff
@@ -0,0 +1,91 @@
+        // to cancel the task after sending it so it doesn't try again.
+
+       let createRequest = Task {
+            do {
```
I think we can remove the async throws in here as well if we're going to just catch any errors inside the function

brendanlensink commented on Netable/Example/Services/NetworkService.swift:
```diff
@@ -0,0 +1,29 @@
+        Task {
+            do {
+                let version = try await netable.request(GetVersionRequest())
+                print(version.buildNumber)
```
If we're going to leave in some print statements, it might be nice to have some formatting and context for them

amyoulton replied:
I opted to remove this because now that I've corrected the fallback resource, it actually hits the "catch" block, which I feel is more significant anyhow as the logging is very clear.

brendanlensink reviewed (commented):
Overall things are looking great! 

Some minor formatting/code stlyle stuff to work on, but great job!

amyoulton commented on Netable/Example/Helpers/CustomDecodingStrategy.swift:
```diff
@@ -0,0 +1,49 @@
+
+    static let convertFromKebabCase = JSONDecoder.KeyDecodingStrategy.custom({ keys in
+        var modifiedKey = ""
+        let key = keys.last!
```
This is actually how the apple docs recommended to do it, so I figured maybe it was actually desirable and I just didn't know the case where it would be! 🙃 I totally thought the exact same thing! 

I'll add in some guard statements to make it more conformant to how we would actually want to do it.

brendanlensink replied:
If it's from the apple docs it's probably fine to just leave it and link to where you got it from

brendanlensink commented:
Another thing I forgot to mention: We should go through the app with concurrency checking set to Complete and try and sort out any warnings it's throwing. Even with minimal it's throwing >10 😬

amyoulton commented on Netable/Example/Views/RootView.swift:
```diff
@@ -0,0 +1,103 @@
+                }
+            }
+            if viewModel.user == nil {
+                loginView
```
Sounds great. I think practically it makes sense to leave the login functionality within the RootVM because of how intertwined it is an immediate it calls the rest of the root functions, but if you think it would be better to separate it I'm happy to do so!

amyoulton commented on Netable/Example/Requests/GraphQL/GetAllPostsQuery.swift:
```diff
@@ -0,0 +1,18 @@
+
+struct GetAllPostsQuery: GraphQLRequest {
+    typealias Parameters = Empty
+    typealias RawResource = SmartUnwrap<[Post]>
```
I opted to change this one to be a SmartUnwrap because they were both using finalize!

amyoulton commented on Netable/Example/MockNetworkingService/JSONData/graphqlPosts.json:
```diff
@@ -0,0 +1,20 @@
+{
```
Had to go ahead and add a second post json file because I changed the original to showcase Lossy, and it doesn't work/cooperate with Graphql from what I can tell!

amyoulton commented on Netable/Example/Models/User.swift:
```diff
@@ -0,0 +1,25 @@
+    let location: String
+    let bio: String
+    let age: Int
+    let loginData: LossyArray<UserLoginData>
```
Wanted to showcase using the LossyArray type as a value within an decodable object. This is working fantastic! (The user json has been updated to contain a new parameter that will fail without lossy).

amyoulton commented on Netable/Example/Requests/GraphQL/UpdatePostsMutation.swift:
```diff
@@ -0,0 +1,35 @@
+    typealias FinalResource = [Post]
+
+    var input: UpdatePostMutationInput
+    var source = GraphQLQuerySource.autoResource
```
I changed this to utilize the `autoResource`!

amyoulton commented:
@brendanlensink Woo! Tons of changes & updates made. 🔥

Here are the significant ones I can think of to report on: 
- updated login to have its own view/VM, which cleans up a lot
- simplified the get user call
- separated the error view
- added in the `.lossy` decoding strategy within the post request
- added a LossyArray to the user object
- updated the post json, the user json, and added an extra json file for the graphql request
- change the graphql get request to use smart unwrap instead
- did a solid walk through every file to check for formatting & cleaned up and refactored. I also temp re-added swift lint to double check!
- lots of updates on the README file - links will need to be updated post-merge as they're currently pointing to this branch.

I did some other small tidies but they've escaped me 😅

brendanlensink commented on Netable/Example/Services/NetworkService.swift:
```diff
@@ -0,0 +1,27 @@
+
+    private let netable = Netable(baseURL: URL(string: "http://localhost:8080/")!)
+
+    func getVersion() {
```
We might consider changing this to be async and bubble the errors up a level to be consistent with our other services

amyoulton replied:
I wrote this one using the Task do demonstrate try/catch within the actual service because that was listed in the project requirements! If we want to be consistent I can definitely change that. We are using try catch within the `createPost` request in the `authNetworkService`!

brendanlensink replied:
okay, lets just leave it. I hadn't intended to specify where users were supposed to use try/catch to handle errors, and our idiomatic solution is usually to pass errors back up to the VM but might as well show both ways

brendanlensink commented on Netable/Example/ViewModels/HomeVM.swift:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    func createPost() {
+        Task { @MainActor in
```
Nitpicky, but does this need to be run on the main thread? AFAICT it doesn't update any UI?

amyoulton replied:
It doesn't and I'll go ahead and remove it. I was initially planning on having it update until I realized that it just wasn't practical given our mock api situation.

amyoulton replied:
It also shouldn't be a try await because it's not an async or throwing function, so I fixed that as well.

brendanlensink commented on README.md:
````diff
@@ -31,11 +31,14 @@ Netable is built on a number of core principles we believe a networking library
 ```swift
 let netable = Netable(baseURL: URL(string: "https://api.thecatapi.com/v1/")!)
 ```
+See [here](#Additional Netable instance parameters) for information on adding additional instance parameters. 
````
I think this [syntax is slightly off](https://gist.github.com/rachelhyman/b1f109155c9dafffe618), it should be #like-this-without-caps

brendanlensink commented on README.md:
```diff
@@ -149,6 +152,35 @@ netable.request(GetCatImages()) { result in
 
+#### Canceling A Request
+
+You're able to easily cancel a request using `.cancel()`, which you can see in action in the [AuthNetworkService](https://github.com/steamclock/netable/blob/aeo/86-new-example-project/Netable/Example/Services/AuthNetworkService.swift#L82) within the Example Project.
```
It might make sense that you need to change your request around a little to retain a reference to it, this might be worth a small example code block

amyoulton replied:
Sorry, I can't quite make sense of this! The request its pointing to uses the cancel by default, unless someone chooses to comment it out. What do you mean by "retain a reference to it"? I can definitely provide a small code block, I just want to clarify this point!

brendanlensink replied:
Woof I'm sorry, I did not proofread that comment well enough 😬 

What I was _trying_ to say is that to use `cancel()`, you need to retain a reference to your Task:

```
 let createRequest = Task {
               let result = try await netable.request()
}

createRequest.cancel()
```

Which isn't normal, so we should probably put a little code sample in

brendanlensink commented on README.md:
```diff
@@ -244,19 +276,75 @@ struct GetUserRequest: Request {
 #### Using `requestFailureDelegate`
 
-See [GlobalRequestFailureDelegate](https://github.com/steamclock/netable/blob/master/Netable/NetableExample/ViewController/RootTabBarController.swift) in the Example project for a more detailed example.
+See [GlobalRequestFailureDelegate](https://github.com/steamclock/netable/blob/aeo/86-new-example-project/Netable/Example/Services/ErrorService.swift) in the Example project for a more detailed example.
```
You'd mentioned it in Slack, but I want to point out we'll need to switch these over to master before merging

amyoulton replied:
So I can swap these right to `master/netable` now? I guess as long as I don't change the files or directories, the paths won't change so the links won't break, correct? That's a dumb question I know 😅

brendanlensink replied:
Yeah, might as well swap them over now

brendanlensink reviewed (approved):
I think we're danger close here @amyoulton! Great job 🎉 🎉🎉🎉🎉

I've left a couple small suggestions that shouldn't require a re-review, and want to give @nbrooke a chance to look through things if he's so inclined, but after both of those are complete lets merge away!

amyoulton commented:
@brendanlensink looking through this PR again after the break apparently let my eyeballs actually work because I seemed to entirely ignore that concurrency checking comment.

So I will need to do that as well and then re-submit, which will I believe will warrant a re-review. Thoughts?

nbrooke commented on Netable/Example/MockNetworkingService/JSONData/login.json:
```diff
@@ -0,0 +1,5 @@
+{
+    "email": "cat@netable.com",
```
Very minor point, but we probably shouldn’t use a potentially real domain that we don’t control and isn’t a known sample domain for something like this. I’d be inclined to just use `example.com` here and anywhere else we are using this right now.

nbrooke commented on Netable/Example/Requests/GraphQL/UpdatePostsMutation.swift:
```diff
@@ -0,0 +1,35 @@
+
+    func finalize(raw: GetAllPostsResponse) async throws -> [Post] {
+        raw.posts
+    }
```
Are we doing a manual finalize here to demonstrate how they work, or because smart unwrap doesn’t support this exact construction? If the former we should maybe add a comment saying that’s the case, if the latter, what is it that prevents it from working here?

amyoulton replied:
The former! I'll go ahead and add a comment above finalize specifying that.

nbrooke reviewed (approved):
Couple very minor comments, but overall look really goof! Approving, even though there are some suggested tweaks, because both of these are minor enough hat they don’t really need a re-review if they are addressed.

amyoulton commented:
@brendanlensink alright, this should be final review 🥳 I've added the changes, removed the example project and updated the links. I have **not** added in Sendable conformance though, because this is turning out to be a lot more of a headache than expected.

Since for the most part, conforming to Sendable will happen outside of my project, I was thinking that once this is merged, I'll rebase #115 and conform to Sendable for Netable on that issue, while also adding the few conformances to the example project at the same time, since they will be so interwoven, and the example project does a great job of really showing where the conformance issues lie!

Let me know your thoughts on this, but either way I believe we should merge this first!

brendanlensink reviewed (approved):
🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀
